### ISSUE TITLE
Own training data is 16-bit grayscale

### ISSUE THREAD

**Issue by SvSz** [ayulockin/deepimageinpainting#3]
I am a beginner to ML. Can you tell if your CNNs work for 16-bit data sets?

**Background:**

**1)** To start with i modified to work " Image_Inpainting_Autoencoder_decoder_approach.ipynb " with the MNIST data set.
I think it could work, although it is s 8-bit 28x28 data set. Although the notebook run through completly I did get unusable results. But it went through, that was OK for me, haha.

Adjustments: 

- In different lines I hashed out any use of the labels (y_train, y_test) because i did not see where they are beneficial in inpainting
example:

```python
 traingen = createAugment(x_train, _)
```

- in __init__:

```python
 __init__ : dim=(28,28), n_channels = 1
```   

instead of 
```python
 __init__ : dim=(32,32), n_channels = 3
```

- __data_generation
```python
res = cv2.bitwise_and(tmp_image, mask).reshape(28,28,1)
```

- unet_like:
```python
inputs = keras.layers.Input((28, 28, 1))
```
and lines from pool3 to conv7 were hashed out, the others were adjusted, because 28x28 cannot be pooled as often as 32x32

- some other minor changes

__________________________________________
**My main Question:**

**2)** After it ran through the single-band (grayscale) MNIST-Dataset I decided to import my own data set.
It consists of 1000 Train images and 100 Test images (to start with).  The source is SRTM1 (Satellite Data) provided by Google Earth Engine. Every Image has 256x256 pixels and is 16-bit grayscale coded. So instead of values from 0 to 255 this data set has now a range from 27 to 3531 pixel-values (so only a small window in the 16-bit range of 2^16 values)

=> Can you tell me whether your notebook works for this 16-bit data set? 

- In createAugment you often used 255 and int8 
example:         

```python
mask = np.full(tmp_image.shape, 255, np.uint8)
        mask[y:y+h,x:x+w] = 0
        res = cv2.bitwise_and(tmp_image, mask)

        X_batch[i,] = res/255
```
        
What can I do here? I think i did not quiet understand the purpose here.

Great Work, looking forward to work with this!

Cheers!

**Comment by sayakpaul** [comment]
Thank you for your interest in our work. 

> Although the notebook run through completly I did get unusable results.

It can happen since the architecture might need a bit of a tweak depending on the data set you are providing to it. 

> Every Image has 256x256 pixels and is 16-bit grayscale coded. So instead of values from 0 to 255 this data set has now a range from 27 to 3531 pixel-values.

Our architecture could be adapted to support this. However, you will need to take care of the pixel scaling appropriately. For int8, in order to get the pixel value range to [0, 1]. This is usually good practice when training neural nets because it helps to avoid isotropic scaling. 

> What can I do here? I think i did not quiet understand the purpose here.

Are you referring to the typecasting? It's there because cv2 won't support anything else. 


> In different lines I hashed out any use of the labels (y_train, y_test)

Yes, the labels aren't used. We kept it to support conditional training but we didn't incorporate that. 

Overall, I would suggest using Partial Convs for your data which is demonstrated [here](https://github.com/ayulockin/deepimageinpainting/blob/master/Image_Inpainting_Partial_Convolution.ipynb).

**Comment by SvSz** [comment]
Thank you very much for your imediate answer!! 

> It can happen since the architecture might need a bit of a tweak depending on the data set you are providing to it.

yes. i was not flaying you. i assumed that i did something wrong. but that was ok :)

>  in order to get the pixel value range to [0, 1]

so you say, there is actually no benefit in using 16-bit data for inpainting? *confused*

> Are you referring to the typecasting? It's there because cv2 won't support anything else.

ok thanks i was not aware of that! 

> Overall, I would suggest using Partial Convs for your data which is demonstrated here.

i will try to adapt it to my (1000,256,256,1) shaped dataset. thanks!

**Comment by sayakpaul** [comment]
> so you say, there is actually no benefit in using 16-bit data for inpainting? confused

No, I am not saying that. No matter which precision you are using a rescaling step or a normalization step is necessary before your data is fed to the network.

**Comment by SvSz** [comment]
Alright! Thank you! I got the normalization process wrong! Now I got it. But I have further questions now, I hope you have time to help :)

**Comment by SvSz** [comment]
I expanded the code from your Partial Convolution like this:

```python
## VT 1000 Dataset
(x_train, x_test) = VT_1000

# min val
minElement_train = np.amin(x_train)
minElement_test = np.amin(x_test)
minElement = np.amin(np.array(minElement_train, minElement_test))


# min val
maxElement_train = np.amax(x_train)
maxElement_test = np.amax(x_test)
maxElement = np.amax(np.array(maxElement_train, maxElement_test))


# shape
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

print('Minimum Value in Dataset:', minElement)
print('Maximum Value in Dataset:', maxElement)
```
VT_1000 is max Dataset

Its output is
```python
>>> x_train shape: (1000, 256, 256)
>>> 1000 train samples
>>> 100 test samples
>>> Minimum Value in Dataset: 49
>>> Maximum Value in Dataset: 3531
```

```python
## Get first 32 images as samples
sample_images = x_train[:32]
#sample_labels = y_train[:32]

fig = plt.figure(figsize=(16., 8.))
grid = ImageGrid(fig, 111,
                 nrows_ncols=(4, 8),  # creates 4x8 grid of axes
                 axes_pad=0.3,  # pad between axes in inch.
                 )

for ax, image in zip(grid, sample_images):
  ax.imshow(image, cmap="gray")

plt.show()
```
Showing my Images: 
![valleysshow](https://user-images.githubusercontent.com/78486927/107081693-612d3a80-67f3-11eb-8537-178e76fe11b6.jpg)



I implemeneted maxElement here instead of 255
changed dim and n_channels and flipped around x1, y1, x2, y2 because that seemed more convenient to me but nvm...
in __createMask i also changed np.unit8 into np.unit16 and i adjusted the shapes to 256

```python

## Ref: https://stanford.edu/~shervine/blog/keras-how-to-generate-data-on-the-fly.
class createAugment(keras.utils.Sequence):
  'Generates data for Keras'
  def __init__(self, X, y, batch_size=32, dim=(256, 256), n_channels=1, shuffle=True):
      'Initialization'
      self.batch_size = batch_size 
      self.X = X 
      self.y = y
      self.dim = dim
      self.n_channels = n_channels
      self.shuffle = shuffle
      
      self.on_epoch_end()

  def __len__(self):
      'Denotes the number of batches per epoch'
      return int(np.floor(len(self.X) / self.batch_size))

  def __getitem__(self, index):
      'Generate one batch of data'
      # Generate indexes of the batch
      indexes = self.indexes[index*self.batch_size:(index+1)*self.batch_size]

      # Generate data
      X_inputs, y_output = self.__data_generation(indexes)
      return X_inputs, y_output

  def on_epoch_end(self):
      'Updates indexes after each epoch'
      self.indexes = np.arange(len(self.X))
      if self.shuffle:
          np.random.shuffle(self.indexes)

  def __data_generation(self, idxs):
    # Masked_images is a matrix of masked images used as input
    Masked_images = np.empty((self.batch_size, self.dim[0], self.dim[1], self.n_channels)) # Masked image
    # Mask_batch is a matrix of binary masks used as input
    Mask_batch = np.empty((self.batch_size, self.dim[0], self.dim[1], self.n_channels)) # Binary Masks
    # y_batch is a matrix of original images used for computing error from reconstructed image
    y_batch = np.empty((self.batch_size, self.dim[0], self.dim[1], self.n_channels)) # Original image
    

    ## Iterate through random indexes
    for i, idx in enumerate(idxs):
      image_copy = self.X[idx].copy()
  
      ## Get mask associated to that image
      masked_image, mask = self.__createMask(image_copy)
      
      Masked_images[i,] = masked_image/maxElement
      Mask_batch[i,] = mask/maxElement
      y_batch[i] = self.y[idx]/maxElement

    ## Return mask as well because partial convolution require the same.
    return [Masked_images, Mask_batch], y_batch

  def __createMask(self, img):
    ## Prepare masking matrix
    mask = np.full((256,256), maxElement, np.uint16) ## White background
    for _ in range(np.random.randint(1, 10)): # define random amount of lines
      # Get random x locations to start line
      x1, y1 = np.random.randint(1, 256), np.random.randint(1, 256)
      # Get random y locations to start line
      x2, y2 = np.random.randint(1, 256), np.random.randint(1, 256)
      # Get random thickness of the line drawn
      thickness = np.random.randint(1, 3)
      # Draw black line on the white mask; (0,0,0) for RGB
      cv2.line(mask,(x1,y1),(x2,y2),(0,0,0),thickness)

    ## Mask the image
    masked_image = img.copy()
    masked_image[mask==0] = maxElement

    return masked_image, mask

```

the rest stayed quiet the same. 

```python
## Prepare training and testing mask-image pair generator
traingen = createAugment(x_train, x_train)
testgen = createAugment(x_test, x_test, shuffle=False)
```

```python

# Legend: Original Image | Mask generated | Masked Image

## Examples
sample_idx = 90 ## Change this to see different batches

[masked_images, masks], nomask = traingen[sample_idx]
sample_images = [None]*(len(masked_images)+len(masks)+len(nomask))
sample_images[::3] = nomask
# masks[]
sample_images[1::3] = masks 
sample_images[2::3] = masked_images 

fig = plt.figure(figsize=(17., 8.))
grid = ImageGrid(fig, 111,  # similar to subplot(111)
                 nrows_ncols=(4, 9),  # creates 4x9 grid of axes
                 axes_pad=0.3,  # pad between axes in inch.
                 )

for ax, image in zip(grid, sample_images):
  ax.imshow(image.reshape(256,256), cmap="gray")

plt.show()

```


Final Output to visualize the masked images:
![black and regular white dots](https://user-images.githubusercontent.com/78486927/107081471-07c50b80-67f3-11eb-84a1-ca752d3f35b9.jpg)


Can you understand what I did wrong? How is it that i only surmise some regular white dots in these black images?

I assume it happens in the  class createAugment(keras.utils.Sequence)  ¯\_(ツ)_/¯

**Comment by SvSz** [comment]
I deleted the n_channels from __init__ and it worked! maxElement is explained above and 256 is the image size.

```python

## Ref: https://stanford.edu/~shervine/blog/keras-how-to-generate-data-on-the-fly.
class createAugment(keras.utils.Sequence):
  'Generates data for Keras'
  def __init__(self, X, y, batch_size=16, dim=(256, 256), shuffle=True):
      'Initialization'
      self.batch_size = batch_size 
      self.X = X 
      self.y = y
      self.dim = dim
      self.shuffle = shuffle
      
      self.on_epoch_end()

  def __len__(self):
      'Denotes the number of batches per epoch'
      return int(np.floor(len(self.X) / self.batch_size))

  def __getitem__(self, index):
      'Generate one batch of data'
      # Generate indexes of the batch
      indexes = self.indexes[index*self.batch_size:(index+1)*self.batch_size]

      # Generate data
      X_inputs, y_output = self.__data_generation(indexes)
      return X_inputs, y_output

  def on_epoch_end(self):
      'Updates indexes after each epoch'
      self.indexes = np.arange(len(self.X))
      if self.shuffle:
          np.random.shuffle(self.indexes)

  def __data_generation(self, idxs):
    # Masked_images is a matrix of masked images used as input
    Masked_images = np.empty((self.batch_size, self.dim[0], self.dim[1])) # Masked image
    # Mask_batch is a matrix of binary masks used as input
    Mask_batch = np.empty((self.batch_size, self.dim[0], self.dim[1])) # Binary Masks
    # y_batch is a matrix of original images used for computing error from reconstructed image
    y_batch = np.empty((self.batch_size, self.dim[0], self.dim[1])) # Original image
    

    ## Iterate through random indexes
    for i, idx in enumerate(idxs):
      image_copy = self.X[idx].copy()
  
      ## Get mask associated to that image
      masked_image, mask = self.__createMask(image_copy)
      
      Masked_images[i,] = masked_image/maxElement
      Mask_batch[i,] = mask/maxElement
      y_batch[i] = self.y[idx]/maxElement

    ## Return mask as well because partial convolution require the same.
    return [Masked_images, Mask_batch], y_batch

  def __createMask(self, img):
    ## Prepare masking matrix
    mask = np.full((256,256), maxElement, np.uint8) ## White background
    for _ in range(np.random.randint(1, 10)):
      # Get random x locations to start line
      x1, x2 = np.random.randint(1, 256), np.random.randint(1, 256)
      # Get random y locations to start line
      y1, y2 = np.random.randint(1, 256), np.random.randint(1, 256)
      # Get random thickness of the line drawn
      thickness = np.random.randint(5, 10)
      # Draw black line on the white mask
      cv2.line(mask,(x1,y1),(x2,y2),(0,0,0),thickness)

    ## Mask the image
    masked_image = img.copy()
    masked_image[mask==0] = maxElement

    return masked_image, mask

``` 

![scratched!](https://user-images.githubusercontent.com/78486927/107370797-bca54400-6ae3-11eb-9050-3dd211b3b27b.jpg)

In the next steps I changed the prepare_model. here is an example:

```python
class InpaintingModel:
  '''
  Build UNET like model for image inpaining task.
  '''
  def prepare_model(self, input_size=(256,256,1)):
```

So the Model starts with: 


   _input_1 (InputLayer)            [(None, 256, 256, 1) 0_                                            


going down to a minimum of


  _encoder_output (PConv2D)        [(None, 16, 16, 32), 18464_     


before decoding again to 256.

But then, training the model does not work, I assume it has to do with the grayscale instead of RGB. so one expected dimension is missing.

![model](https://user-images.githubusercontent.com/78486927/107373266-9fbe4000-6ae6-11eb-82b2-e33ecd8c42ca.jpg)






### Is there a way to get this running on 1-scale images? by enwrapping the gray values into the 4th dimension, or by setting requiered dimensons to 3? I don't know where this can be done.

**Comment by SvSz** [comment]
With a 4-dimensional shpae of x_train being (1000,256,256,1) and x_test (100,256,256,1) the fit.generator can now run through!
The next problem I have now is not part of the main topic here:  All values of the inpainted_image are the same

PS: there is a typo in the script namely iMpainted_image instead of iNpainted_image.

**Comment by SvSz** [comment]
With Cifar-10 and MNIST's numbers working, I assume my data to be too smooth for this CNN. It won't work because of the lack of edges and structure. Please confirm or prove the opposite.

Back to the topic: 

To get 16-bit working, you need to normalize it like so:

```python
x_test = (x_testO - np.amin(x_trainO))/(np.amax(x_trainO) - np.amin(x_trainO))
x_train = (x_trainO - np.amin(x_trainO))/(np.amax(x_trainO) - np.amin(x_trainO))
```

**Comment by SvSz** [comment]
I turned my Dataset into the size of 32,32,1 tiles instead of 256,256,1
It returned with success and gave me this: 

## Legend: Original Image w/ Mask | Mask generated | Inpainted Image | Ground Truth

![eureka2](https://user-images.githubusercontent.com/78486927/109011279-03349a00-76b1-11eb-8f7f-3291043185c7.jpg)

(Different brightness values from original to inpainted image results from issues displaying the similar values equally. python decodes similar values differently in different images. I think it results from the original images being 16-bit.)

Great work guys! Problem Solved!

Close at your wish!

**Comment by sayakpaul** [comment]
Great. Thanks for working through this and apologies for not being able to get back timely.
